go/cmd/vpnkit-forwarder/many.go

```diff
@@ -46,72 +45,13 @@ func manyPorts() {
 					log.Printf("Error accepting connection: %#v", err)
 					return // no more listening
 				}
-				go func(conn net.Conn) {
-					// Read header which describes TCP/UDP and destination IP:port
-					d, err := unmarshalDestination(conn)
-					if err != nil {
-						log.Printf("Failed to unmarshal header: %#v", err)
-						conn.Close()
-						return
-					}
-					switch d.Proto {
-					case TCP:
-						backendAddr := net.TCPAddr{IP: d.IP, Port: int(d.Port), Zone: ""}
-						libproxy.HandleTCPConnection(conn.(libproxy.Conn), &backendAddr, quit)
-						break
-					case UDP:
-						backendAddr := &net.UDPAddr{IP: d.IP, Port: int(d.Port), Zone: ""}
-
-						proxy, err := libproxy.NewUDPProxy(backendAddr, libproxy.NewUDPConn(conn), backendAddr)
-						if err != nil {
-							log.Printf("Failed to setup UDP proxy for %s: %#v", backendAddr, err)
-							conn.Close()
-							return
-						}
-						proxy.Run()
-						break
-					default:
-						log.Printf("Unknown protocol: %d", d.Proto)
-						conn.Close()
-						return
-					}
-				}(conn)
+				if err := libproxy.HandleMultiplexedConnections(conn, quit); err != nil {
+					log.Println(err)
+					conn.Close()
```

Comment:
From what I can see, we only close on errors, but I wonder whether we should have `defer conn.Close()`, but I don't know enough about this.

Reply:
It's not obvious (which is bad) but I think `HandleMultiplexedConnections` calls `HandleTCPConnection` which calls `Close` https://github.com/moby/vpnkit/blob/e1e5e6b8a0b0dea4e82dd5d23f4f8eb49f5ec69c/go/pkg/libproxy/tcp_proxy.go#L81 and `UDPProxy.Run` which ... looks like it might leak the listening socket? It's possible we didn't notice because in the one-process-per-port model it would be cleaned up anyway when the process quits.

I think `HandleTCPConnection` should probably not call `client.Close()`, since `client` is passed in from the caller. I think it would be clearer if the caller retained ownership of the connection and called `defer conn.Close()` as you suggest.

Reply:
Closing on error case is obvious from outside the library call, but I'm not so sure how to handle the two non-error cases of `client.Close()` in `HandleTCPConnection`...